docs/source/step-by-step.rst

```diff
@@ -1,15 +1,760 @@
+    ---------+--------+--------+---------+--------------
+        0    |    72M | 124418  |   577   |     71859761
+        1    |     0  |     0  |   n/a   |            0
+        2    |     0  |     0  |   n/a   |            0
+        3    |     0  |     0  |   n/a   |            0
+        4    |     0  |     0  |   n/a   |            0
+        5    |     0  |     0  |   n/a   |            0
+        6    |     0  |     0  |   n/a   |            0
+        7    |     0  |     0  |   n/a   |            0
+       136M (   135730786) cycles total
+
+Another optimization we can do is called "loop unrolling". Because
+``input_depth`` is always a multiple of 4, we can make the innermost loop do
+4 times as much work before checking the loop conditions. Implementing this
+change should make your innermost loop look like:
+
+.. code-block:: C++
+
+    for (int in_channel = 0; in_channel < input_depth; in_channel += 4) {
+      int32_t input_val = input_data[Offset(input_shape, batch, in_y,
+                                              in_x, in_channel)];
+      int32_t filter_val = filter_data[Offset(
+          filter_shape, out_channel, filter_y, filter_x, in_channel)];
+      acc += filter_val * (input_val + 128);
+      
+      input_val = input_data[Offset(input_shape, batch, in_y,
+                                              in_x, in_channel + 1)];
+      filter_val = filter_data[Offset(
+          filter_shape, out_channel, filter_y, filter_x, in_channel + 1)];
+      acc += filter_val * (input_val + 128);
+  
+      input_val = input_data[Offset(input_shape, batch, in_y,
+                                              in_x, in_channel + 2)];
+      filter_val = filter_data[Offset(
+          filter_shape, out_channel, filter_y, filter_x, in_channel + 2)];
+      acc += filter_val * (input_val + 128);
+  
+      input_val = input_data[Offset(input_shape, batch, in_y,
+                                              in_x, in_channel + 3)];
+      filter_val = filter_data[Offset(
+          filter_shape, out_channel, filter_y, filter_x, in_channel + 3)];
+      acc += filter_val * (input_val + 128);
+    }
+
+After this change we get another significant speed-up in our golden tests:
+
+::
+    
+     Counter |  Total | Starts | Average |     Raw
+    ---------+--------+--------+---------+--------------
+        0    |    54M | 124418  |   431   |     53743879
+        1    |     0  |     0  |   n/a   |            0
+        2    |     0  |     0  |   n/a   |            0
+        3    |     0  |     0  |   n/a   |            0
+        4    |     0  |     0  |   n/a   |            0
+        5    |     0  |     0  |   n/a   |            0
+        6    |     0  |     0  |   n/a   |            0
+        7    |     0  |     0  |   n/a   |            0
+       117M (   117297894) cycles total
+
+Even with the simplest possible software specialization, we've already seen
+massive gains in performance.
```

Comment:
Do the maths here to quantify "massive":
 - in this loop 54M / 113 = 0.48 => 52% reduction in time spent
 - in total 117M/183M = 0.64 => 36% reduction in time spent.


Also (maybe for a future PR), encourage the reader to record progress (in a spreadsheet or a notebook)